### PR TITLE
Remove the use of registry in fake clientset

### PR DESCRIPTION
**Pull request by caesarxuchao**
Addressing https://github.com/kubernetes/kubernetes/pull/42095#issuecomment-298136406

The fake clientset needed registry for multiple reasons, and this PR removed these needs:
1. translate GroupVersionResource to GroupVersionKind, the internal `tracker` (a fake apiserver) stores the objects in a map and the key is gvk. This PR changes the map key to be gvr.
2. client-go/testing/fixture.go had a function `checkNamespaced`, which used the registry as the ground truth to verify if the request send by client is properly namespace scoped. Every unit test that used the generated fake clientset would eventually call this function to verify the request sent by the fake clietnset. Such a check is useless, this PR removed `checkNamespaced` completely.
3. When the `tracker` reacted to a fake List request, it asked the registry to translate GVR to GVK, and created a list of type `GVK+"List"`. This PR lets the ListActionImpl specify the GVK explicitly to get rid of the need of the GVR to GVK translation.

(the first commit is #45180)

### CODE REVIEW

**Comment by caesarxuchao** [comment]
Rebased on top of #45180.

**Comment by caesarxuchao** [comment]
@sttts @deads2k could you help review this change? You can ignore the commit "move metrics to staging", which is #45180, and you can ignore most of the commit "generated clientset changes", which are all generated code. Thanks.

**Comment by caesarxuchao** [comment]
cc @lavalamp

File: staging/src/k8s.io/client-go/testing/fixture.go [hunk 1]
```diff
@@ -157,19 +142,24 @@ var _ ObjectTracker = &tracker{}
...
 	}
 }
 
-func (t *tracker) List(gvk schema.GroupVersionKind, ns string) (runtime.Object, error) {
+func (t *tracker) List(gvr schema.GroupVersionResource, gvk schema.GroupVersionKind, ns string) (runtime.Object, error) {
 	// Heuristic for list kind: original kind + List suffix. Might
 	// not always be true but this tracker has a pretty limited
 	// understanding of the actual API model.
+	if gvr.Version == "" {
+
```
Comment (sttts) [inline staging/src/k8s.io/client-go/testing/fixture.go]: something missing?

File: staging/src/k8s.io/client-go/testing/fixture.go [hunk 2]
```diff
@@ -157,19 +142,24 @@ var _ ObjectTracker = &tracker{}
...
+	}
 	listGVK := gvk
 	listGVK.Kind = listGVK.Kind + "List"
+	if listGVK.Version == "" {
+		listGVK.Version = runtime.APIVersionInternal
```
Comment (sttts) [inline staging/src/k8s.io/client-go/testing/fixture.go]: why is this needed? If it is, please add a comment why.

File: staging/src/k8s.io/client-go/testing/fixture.go [hunk 3]
```diff
@@ -251,90 +237,95 @@ func (t *tracker) Add(obj runtime.Object) error {
...
 	if len(gvks) == 0 {
 		return fmt.Errorf("no registered kinds for %v", obj)
 	}
-
-	t.lock.Lock()
-	defer t.lock.Unlock()
-
 	for _, gvk := range gvks {
-		gr := schema.GroupResource{Group: gvk.Group, Resource: gvk.Kind}
+		gvr, _ := meta.KindToResource(gvk)
```
Comment (sttts) [inline staging/src/k8s.io/client-go/testing/fixture.go]: Please add a comment that this function is best-effort, using some heuristics only.

Comment (sttts) [inline staging/src/k8s.io/client-go/testing/fixture.go]: What happens below if it guesses wrong?

File: pkg/kubectl/cmd/util/factory_object_mapping_test.go [hunk 4]
```diff
@@ -20,6 +20,8 @@ import (
 	"reflect"
 	"testing"
 
+	"time"
```
Comment (sttts) [inline pkg/kubectl/cmd/util/factory_object_mapping_test.go]: nit: no empty line above.

File: staging/src/k8s.io/client-go/testing/actions.go [hunk 5]
```diff
@@ -47,20 +47,22 @@ func NewGetAction(resource schema.GroupVersionResource, namespace, name string)
...
 	action.Resource = resource
+	action.Kind = kind
 	labelSelector, fieldSelector, _ := ExtractFromListOptions(opts)
 	action.ListRestrictions = ListRestrictions{labelSelector, fieldSelector}
 
 	return action
 }
 
-func NewListAction(resource schema.GroupVersionResource, namespace string, opts interface{}) ListActionImpl {
+func NewListAction(resource schema.GroupVersionResource, kind schema.GroupVersionKind, namespace string, opts interface{}) ListActionImpl {
```
Comment (sttts) [inline staging/src/k8s.io/client-go/testing/actions.go]: are both gv's always equal?

File: staging/src/k8s.io/client-go/testing/fixture.go [hunk 6]
```diff
@@ -237,7 +223,7 @@ func (t *tracker) Get(gvk schema.GroupVersionKind, ns, name string) (runtime.Obj
 
 	if status, ok := obj.(*metav1.Status); ok {
 		if status.Details != nil {
-			status.Details.Kind = gvk.Kind
+			status.Details.Kind = gvr.Resource
```
Comment (sttts) [inline staging/src/k8s.io/client-go/testing/fixture.go]: please comment why using a resource name here is good enough

File: staging/src/k8s.io/client-go/testing/fixture.go (hunk 3 above)
Comment (caesarxuchao) [inline staging/src/k8s.io/client-go/testing/fixture.go]: This is the same function used to construct a restmapper, so it's accurate. GVK is only mapped to a single GVR.

File: staging/src/k8s.io/client-go/testing/fixture.go (hunk 2 above)
Comment (caesarxuchao) [inline staging/src/k8s.io/client-go/testing/fixture.go]: Added.

File: staging/src/k8s.io/client-go/testing/fixture.go (hunk 1 above)
Comment (caesarxuchao) [inline staging/src/k8s.io/client-go/testing/fixture.go]: Removed. I was hesitating whether to set internal gvr version to empty or to "__internal".

File: pkg/kubectl/cmd/util/factory_object_mapping_test.go (hunk 4 above)
Comment (caesarxuchao) [inline pkg/kubectl/cmd/util/factory_object_mapping_test.go]: Done.

File: staging/src/k8s.io/client-go/testing/actions.go (hunk 5 above)
Comment (caesarxuchao) [inline staging/src/k8s.io/client-go/testing/actions.go]: They are equal in all current use cases.

File: staging/src/k8s.io/client-go/testing/fixture.go (hunk 3 above)
Comment (deads2k) [inline staging/src/k8s.io/client-go/testing/fixture.go]: > This is the same function used to construct a restmapper, so it's accurate. GVK is only mapped to a single GVR.

It's not.  This is a heuristic match and the default for building one from types.  The actual registration can specify what it wants and the client builds it based on discovery.

File: staging/src/k8s.io/client-go/testing/fixture.go (hunk 6 above)
Comment (caesarxuchao) [inline staging/src/k8s.io/client-go/testing/fixture.go]: This looks like dead code. I removed the `if` block and no tests were broken. I'll just remove it.

File: staging/src/k8s.io/client-go/testing/fixture.go (hunk 3 above)
Comment (caesarxuchao) [inline staging/src/k8s.io/client-go/testing/fixture.go]: I see. If a tests uses such resources, they cannot initialize the tracker with pre-existing objects via the `tracker.Add()` function, they'll need to call the `Create()` function of the fake clientset. I'll add to comment.

**Comment by caesarxuchao** [comment]
@sttts @deads2k comments addressed, PTAL. Thanks.

File: staging/src/k8s.io/client-go/testing/fixture.go [hunk 7]
```diff
@@ -144,32 +129,35 @@ func ObjectReaction(tracker ObjectTracker, mapper meta.RESTMapper) ReactionFunc
...
-func (t *tracker) List(gvk schema.GroupVersionKind, ns string) (runtime.Object, error) {
+func (t *tracker) List(gvr schema.GroupVersionResource, gvk schema.GroupVersionKind, ns string) (runtime.Object, error) {
 	// Heuristic for list kind: original kind + List suffix. Might
 	// not always be true but this tracker has a pretty limited
 	// understanding of the actual API model.
 	listGVK := gvk
 	listGVK.Kind = listGVK.Kind + "List"
+	// GVK does have the concept of "internal version". The scheme recognizes
+	// the runtime.APIVersionInternal, but not the empty string.
+	if listGVK.Version == "" {
```
Comment (sttts) [inline staging/src/k8s.io/client-go/testing/fixture.go]: where is this actually called with an empty version? A gvk without a version looks wrong.

File: staging/src/k8s.io/client-go/testing/fixture.go [hunk 8]
```diff
@@ -251,90 +232,95 @@ func (t *tracker) Add(obj runtime.Object) error {
...
 	for _, gvk := range gvks {
-		gr := schema.GroupResource{Group: gvk.Group, Resource: gvk.Kind}
+		// NOTE: KindToResource is a heuristic and default match. The actual
+		// registration in apiserver can specify arbitrary route for a gvk. If a
+		// test uses such objects, it cannot preset the tracker with objects via
+		// Add(). Instead, it should trigger the Create() function of the
+		// tracker, where an arbitrary gvr can be specified.
+		gvr, _ := meta.KindToResource(gvk)
+		// Resource doesn't have the concept of "__internal" version, just set it to "".
+		if gvr.Version == runtime.APIVersionInternal {
```
Comment (sttts) [inline staging/src/k8s.io/client-go/testing/fixture.go]: this version mangling is ugly. I haven't followed all the code paths to check whether it's necessary. My gut feeling: let's save internal objects with internal gvr.

Comment (caesarxuchao) [inline staging/src/k8s.io/client-go/testing/fixture.go]: That means we need to change hundreds lines of manually written test code like this [one](https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/deployment/deployment_controller_test.go#L163). It also requires changing the generated fake clients.

Apart from the massive changes, i don't feel right writing `schema.GroupVersionResource{Group:"core, Version:"__internal", Resource: "Pod"}` in tests because there is no such GVR in production code.

File: staging/src/k8s.io/client-go/testing/fixture.go (hunk 7 above)
Comment (caesarxuchao) [inline staging/src/k8s.io/client-go/testing/fixture.go]: Mostly in the generated fake clients, like this [line](https://github.com/kubernetes/kubernetes/pull/45159/files#diff-1ce5095abadb94bcbff3015537351389R37). These are hidden from unit test writers so i think we should go with the approach that causes minimal changes.

Comment (caesarxuchao) [inline staging/src/k8s.io/client-go/testing/fixture.go]: Mostly in the generated fake clients, like this [line](https://github.com/kubernetes/kubernetes/pull/45159/files#diff-1ce5095abadb94bcbff3015537351389R37). These are hidden from unit test writers so i think we should go with the approach that causes minimal changes.

File: staging/src/k8s.io/client-go/testing/fixture.go (hunk 8 above)
Comment (sttts) [inline staging/src/k8s.io/client-go/testing/fixture.go]: Got it, makes sense.

**Comment by sttts** [comment]
lgtm after squashing.

**Comment by caesarxuchao** [comment]
Squashed to two commits: one for manual changes and one for generated changes.

Applying the lgtm per sttts's comment.

Thanks for reviewing!

**Comment by caesarxuchao** [comment]
@lavalamp could you help approve the changes?

**Comment by caesarxuchao** [comment]
After rebase, the fake custom_metrics client is breaking the unit tests. It's using client-go/api.Regsitry(). I'll fix it and ask @DirectXMan12 to take a look at the new changes.

File: staging/src/k8s.io/metrics/pkg/client/custom_metrics/fake/fake_client.go [hunk 9]
```diff
@@ -52,13 +52,12 @@ func (i GetForActionImpl) GetSubresource() string {
 }
 
 func NewGetForAction(groupKind schema.GroupKind, namespace, name string, metricName string, labelSelector labels.Selector) GetForActionImpl {
-	mapping, err := api.Registry.RESTMapper().RESTMapping(groupKind)
-	if err != nil {
-		panic(fmt.Sprintf("unable to get REST mapping for groupKind %s while building GetFor action: %v", groupKind.String, err))
-	}
+	// the version doesn't matter
+	gvk := groupKind.WithVersion("")
+	gvr, _ := meta.UnsafeGuessKindToResource(gvk)
```
Comment (caesarxuchao) [inline staging/src/k8s.io/metrics/pkg/client/custom_metrics/fake/fake_client.go]: client-go/kubernetes/import_known_versions.go was recently removed so client-go/api.Registry.RESTMapper() doesn't work anymore.
Using the client-go/api.Registry is the same as calling the UnsafeGuessKindToResource method, both are statically built.

Comment (caesarxuchao) [inline staging/src/k8s.io/metrics/pkg/client/custom_metrics/fake/fake_client.go]: @DirectXMan12 could you take a look at the last commit that fixed the fake client of custom_metrics? Thanks.

Comment (DirectXMan12) [inline staging/src/k8s.io/metrics/pkg/client/custom_metrics/fake/fake_client.go]: That seems reasonable, especially since it's just in the fake client.

Comment (caesarxuchao) [inline staging/src/k8s.io/metrics/pkg/client/custom_metrics/fake/fake_client.go]: Thanks.

**Comment by caesarxuchao** [comment]
Squashed the changes and reapply the lgtm.
